internal/table/table.go

```diff
@@ -215,71 +282,42 @@ func (t *Table) Front(ctx context.Context, instanceID uint32) (*Job, error) {
-// and next_run matches it. This is also the behavior when enabling a job.
-// If this is a retry, next_run is set to now+duration and next_sched is not adjusted.
+// With FixedIntervals enabled, next_sched is advanced by the value of interval
+// until it's in the future and next_run matches it.
 //
-// If UpdateNextRun is called on a disabled job without doEnable, as many happen when a job has been
-// disabled during execution, next_sched will advance but next_run will remain NULL.
-// Reversely, if doEnable is true but the job has a non-NULL next_run, the method call is a no-op.
-func (t *Table) UpdateNextRun(ctx context.Context, key Key, delay duration.Duration, randFactor float64, isRetry bool, doEnable bool) error {
-	if isRetry && doEnable {
-		return errors.New("isRetry and doEnable are mutually exclusive parameters")
-	}
-	tx, err := t.db.BeginTx(ctx, nil)
+// If this is a retry, next_run is set to a value based on retry paramters and next_sched is not adjusted.
+//
+// If UpdateNextRun is called on a disabled job, as many happen when a job has
+// been disabled during execution, next_sched will advance but next_run will remain NULL.
+func (t *Table) UpdateNextRun(ctx context.Context, key Key, randFactor float64, retryParams *retry.Retry) error {
+	tx, j, now, err := t.getForUpdate(ctx, key.Path, key.Body)
 	if err != nil {
 		return err
 	}
 	defer tx.Rollback() // nolint: errcheck
-	var locationName string
-	var now, nextSched time.Time
-	var nextRun sql.NullTime
-	s := "SELECT location, next_run, next_sched, IFNULL(CAST(? as DATETIME), UTC_TIMESTAMP())" +
-		" FROM " + t.name + " WHERE path = ? AND body = ?"
-	row := tx.QueryRowContext(ctx, s, t.Clk.NowUTC(), key.Path, key.Body)
-	if err := row.Scan(&locationName, &nextRun, &nextSched, &now); err != nil {
-		return fmt.Errorf("failed to get last run of job: %w", err)
-	}
-	if nextRun.Valid && doEnable {
-		return nil
-	}
-	if locationName == "" {
-		locationName = time.UTC.String() // Default to UTC in case it's omitted somehow in the database.
-	}
-	location, err := time.LoadLocation(locationName)
-	if err != nil {
-		return err
-	}
-	now = now.In(location)
-	nextSched = nextSched.In(location)
 	switch {
-	case !t.FixedIntervals && !doEnable:
-		nextSched = delay.Shift(now)
-		if randFactor > 0 {
-			diff := randomize(nextSched.Sub(now), randFactor)
-			nextSched = now.Add(diff)
+	case retryParams != nil:
+		if j.NextRun.Valid {
+			j.NextRun.Time = retryParams.NextRun(j.NextSched, now)
+		}
+	case t.FixedIntervals:
+		for j.NextSched.Before(now) {
+			j.NextSched = j.Interval.Shift(j.NextSched)
 		}
-	case t.FixedIntervals && !isRetry:
-		for nextSched.Before(now) {
-			nextSched = delay.Shift(nextSched)
+		j.NextRun.Time = j.NextSched
+	case !t.FixedIntervals:
+		j.NextSched = j.Interval.Shift(now)
+		if randFactor > 0 {
+			diff := randomize(j.NextSched.Sub(now), randFactor)
+			j.NextSched = now.Add(diff)
 		}
+		j.NextRun.Time = j.NextSched
```

Comment:
Would be good to have tests to verify all this behavior...

Huh, I mean there's the test `Generic rescheduling won't re-enable a job` which I would think would be blowing up right now?

Reply:
Oh wow, I just got it.  Because `j.NextRun.Valid == false` it actually doesn't matter if you set `j.NextRun.Time = j.NextSched`, it will still insert `NULL` now that you're passing `j.NextRun` instead of `j.NextRun.Time` into the `ExecContext`.

So everything works fine, even if it's perhaps a little subtle.  I think you can actually remove the `if j.NextRun.Valid` check in the `retryParams != nil` case, if you like.

Otherwise, LGTM. 👍

Reply:
I put `j.NextRun.Valid` check to avoid unnecessary calculation in `Retry.NextRun()` but I think it confuses the reader. I will remove that.